src/replication.c

```diff
@@ -3715,78 +3715,72 @@ static void rdbChannelFullSyncWithMaster(connection *conn) {
     rdbChannelAbort();
 }
 
+void replDataBufInit(replDataBuf *buf) {
+    serverAssert(buf->blocks == NULL);
+    buf->size = 0;
+    buf->used = 0;
+    buf->last_num_blocks = 0;
+    buf->mem_used = 0;
+    buf->blocks = listCreate();
+    buf->blocks->free = zfree;
+}
+
+void replDataBufFree(replDataBuf *buf) {
+    if (buf->blocks) listRelease(buf->blocks);
+    buf->blocks = NULL;
+    buf->size = 0;
+    buf->used = 0;
+    buf->last_num_blocks = 0;
+    buf->mem_used = 0;
+}
+
 /* Replication: Replica side.
  * Initialize replica's local replication buffer to accumulate repl stream
  * during rdb channel sync. */
 static void rdbChannelReplDataBufInit(void) {
-    serverAssert(server.repl_full_sync_buffer.blocks == NULL);
-    server.repl_full_sync_buffer.size = 0;
-    server.repl_full_sync_buffer.used = 0;
-    server.repl_full_sync_buffer.last_num_blocks = 0;
-    server.repl_full_sync_buffer.mem_used = 0;
-    server.repl_full_sync_buffer.blocks = listCreate();
-    server.repl_full_sync_buffer.blocks->free = zfree;
+    replDataBufInit(&server.repl_full_sync_buffer);
 }
 
 /* Replication: Replica side.
  * Free replica's local replication buffer */
 static void rdbChannelReplDataBufFree(void) {
-    listRelease(server.repl_full_sync_buffer.blocks);
-    server.repl_full_sync_buffer.blocks = NULL;
-    server.repl_full_sync_buffer.size = 0;
-    server.repl_full_sync_buffer.used = 0;
-    server.repl_full_sync_buffer.last_num_blocks = 0;
-    server.repl_full_sync_buffer.mem_used = 0;
+    replDataBufFree(&server.repl_full_sync_buffer);
 }
 
-/* Replication: Replica side.
- * Reads replication data from master connection into the repl buffer block */
-int rdbChannelReadIntoBuf(connection *conn, replDataBufBlock *b) {
+/* Generic function to read data from connection into a buffer block. */
+static int replDataBufReadIntoBlock(connection *conn,  replDataBuf *buf,
+                replDataBufBlock *block, void (*error_handler)(connection *conn))
+{
     atomicIncr(server.stat_io_reads_processed[IOTHREAD_MAIN_THREAD_ID], 1);
 
-    int nread = connRead(conn, b->buf + b->used, b->size - b->used);
+    int nread = connRead(conn, block->buf + block->used, block->size - block->used);
     if (nread <= 0) {
         if (nread == 0 || connGetState(conn) != CONN_STATE_CONNECTED) {
-            serverLog(LL_WARNING, "Main channel error while reading from master: %s",
-                      connGetLastError(conn));
-            cancelReplicationHandshake(1);
+            error_handler(conn);
         }
         return -1;
     }
 
-    b->used += nread;
-    server.repl_full_sync_buffer.used += nread;
+    block->used += nread;
+    if (buf) buf->used += nread;
```

Comment:
why do we pass `buf` and `block` seperately? `block` is just the tail node of the `buf` right? Also, is there a scenario that `buf` can be null?